File: ts/indicators/marc-chaikin/advance-decline-oscillator/advance-decline-oscillator.ts
import { Bar } from '../../../entities/bar';
import { Scalar } from '../../../entities/scalar';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { ExponentialMovingAverage } from '../../common/exponential-moving-average/exponential-moving-average';
import { SimpleMovingAverage } from '../../common/simple-moving-average/simple-moving-average';
import { AdvanceDeclineOscillatorOutput } from './advance-decline-oscillator-output';
import { AdvanceDeclineOscillatorParams, MovingAverageType } from './advance-decline-oscillator-params';

/** Interface for an indicator that accepts a scalar and returns a value. */
interface LineUpdater {
  update(sample: number): number;
  isPrimed(): boolean;
}

/** Function to calculate mnemonic of an __AdvanceDeclineOscillator__ indicator. */
export const advanceDeclineOscillatorMnemonic = (params: AdvanceDeclineOscillatorParams): string => {
  const maLabel = params.movingAverageType === MovingAverageType.SMA ? 'SMA' : 'EMA';

  return `adosc(${maLabel}${params.fastLength}/${maLabel}${params.slowLength})`;
};

/**
 * AdvanceDeclineOscillator is Marc Chaikin's Advance-Decline (A/D) Oscillator.
 *
 * The Chaikin Oscillator is the difference between a fast and slow moving average
 * of the Accumulation/Distribution Line. It is used to anticipate changes in the A/D Line
 * by measuring the momentum behind accumulation/distribution movements.
 *
 * The value is calculated as:
 *
 *   CLV = ((Close - Low) - (High - Close)) / (High - Low)
 *   AD  = AD_prev + CLV × Volume
 *   ADOSC = FastMA(AD) - SlowMA(AD)
 *
 * When High equals Low, the A/D value is unchanged (no division by zero).
 *
 * Reference:
 *
 * Chaikin, Marc. "Chaikin Oscillator".
 */
export class AdvanceDeclineOscillator extends LineIndicator {
  private ad: number;
  private fastMA: LineUpdater;
  private slowMA: LineUpdater;
  private value: number;

  /**
   * Constructs an instance given the parameters.
   */
  public constructor(params: AdvanceDeclineOscillatorParams) {
    super();

    const fastLength = Math.floor(params.fastLength);
    const slowLength = Math.floor(params.slowLength);

    if (fastLength < 2) {
      throw new Error('fast length should be greater than 1');
    }

    if (slowLength < 2) {
      throw new Error('slow length should be greater than 1');
    }

    this.mnemonic = advanceDeclineOscillatorMnemonic(params);
    this.description = 'Chaikin Advance-Decline Oscillator ' + this.mnemonic;

    this.ad = 0;

    if (params.movingAverageType === MovingAverageType.SMA) {
      this.fastMA = new SimpleMovingAverage({ length: fastLength });
      this.slowMA = new SimpleMovingAverage({ length: slowLength });
    } else {
      this.fastMA = new ExponentialMovingAverage({
        length: fastLength,
        firstIsAverage: params.firstIsAverage ?? false,
      });
      this.slowMA = new ExponentialMovingAverage({
        length: slowLength,
        firstIsAverage: params.firstIsAverage ?? false,
      });
    }

    this.value = Number.NaN;
    this.primed = false;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.AdvanceDeclineOscillator,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: AdvanceDeclineOscillatorOutput.AdvanceDeclineOscillatorValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the indicator with the given sample (H=L=C, volume=1, so AD is unchanged). */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return Number.NaN;
    }

    return this.updateHLCV(sample, sample, sample, 1);
  }

  /** Updates the indicator with the given high, low, close, and volume values. */
  public updateHLCV(high: number, low: number, close: number, volume: number): number {
    if (Number.isNaN(high) || Number.isNaN(low) || Number.isNaN(close) || Number.isNaN(volume)) {
      return Number.NaN;
    }

    // Compute cumulative AD.
    const tmp = high - low;
    if (tmp > 0) {
      this.ad += ((close - low) - (high - close)) / tmp * volume;
    }

    // Feed AD to both MAs.
    const fast = this.fastMA.update(this.ad);
    const slow = this.slowMA.update(this.ad);
    this.primed = this.fastMA.isPrimed() && this.slowMA.isPrimed();

    if (Number.isNaN(fast) || Number.isNaN(slow)) {
      this.value = Number.NaN;
      return this.value;
    }

    this.value = fast - slow;
    return this.value;
  }

  /** Updates the indicator given the next bar sample, extracting HLCV. */
  public override updateBar(sample: Bar): IndicatorOutput {
    const v = this.updateHLCV(sample.high, sample.low, sample.close, sample.volume);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return [scalar];
  }
}

File: go/indicators/marcchaikin/advancedeclineoscillator/advancedeclineoscillator.go
package advancedeclineoscillator

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/common/exponentialmovingaverage"
	"zpano/indicators/common/simplemovingaverage"
	"zpano/indicators/core"
)

// lineUpdater is an interface for indicators that accept a single scalar and return a value.
type lineUpdater interface {
	Update(float64) float64
	IsPrimed() bool
}

// AdvanceDeclineOscillator is Marc Chaikin's Advance-Decline (A/D) Oscillator.
//
// The Chaikin Oscillator is the difference between a fast and slow moving average
// of the Accumulation/Distribution Line. It is used to anticipate changes in the A/D Line
// by measuring the momentum behind accumulation/distribution movements.
//
// The value is calculated as:
//
//	CLV = ((Close - Low) - (High - Close)) / (High - Low)
//	AD  = AD_previous + CLV × Volume
//	ADOSC = FastMA(AD) - SlowMA(AD)
//
// When High equals Low, the A/D value is unchanged (no division by zero).
//
// Reference:
//
// Chaikin, Marc. "Chaikin Oscillator".
type AdvanceDeclineOscillator struct {
	mu sync.RWMutex
	core.LineIndicator
	ad     float64
	fastMA lineUpdater
	slowMA lineUpdater
	value  float64
	primed bool
}

// NewAdvanceDeclineOscillator returns an instance of the indicator created using supplied parameters.
func NewAdvanceDeclineOscillator(p *AdvanceDeclineOscillatorParams) (*AdvanceDeclineOscillator, error) {
	const (
		invalid   = "invalid advance-decline oscillator parameters"
		fmts      = "%s: %s"
		fmtw      = "%s: %w"
		minLength = 2
	)

	if p.FastLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "fast length should be greater than 1")
	}

	if p.SlowLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "slow length should be greater than 1")
	}

	var fastMA, slowMA lineUpdater

	var maLabel string

	switch p.MovingAverageType {
	case SMA:
		maLabel = "SMA"

		fast, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: p.FastLength})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		slow, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: p.SlowLength})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		fastMA = fast
		slowMA = slow
	default:
		maLabel = "EMA"

		fast, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         p.FastLength,
				FirstIsAverage: p.FirstIsAverage,
			})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		slow, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         p.SlowLength,
				FirstIsAverage: p.FirstIsAverage,
			})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		fastMA = fast
		slowMA = slow
	}

	mnemonic := fmt.Sprintf("adosc(%s%d/%s%d)", maLabel, p.FastLength, maLabel, p.SlowLength)
	desc := "Chaikin Advance-Decline Oscillator " + mnemonic

	barFunc, _ := entities.BarComponentFunc(entities.BarClosePrice)
	quoteFunc, _ := entities.QuoteComponentFunc(entities.DefaultQuoteComponent)
	tradeFunc, _ := entities.TradeComponentFunc(entities.DefaultTradeComponent)

	a := &AdvanceDeclineOscillator{
		fastMA: fastMA,
		slowMA: slowMA,
		value:  math.NaN(),
	}

	a.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, a.Update)

	return a, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *AdvanceDeclineOscillator) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *AdvanceDeclineOscillator) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.AdvanceDeclineOscillator,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Update updates the indicator with the given sample.
// Since scalar updates use the same value for H, L, C, the range is 0 and AD is unchanged,
// but the unchanged AD value is still fed to the MAs.
func (s *AdvanceDeclineOscillator) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return math.NaN()
	}

	return s.UpdateHLCV(sample, sample, sample, 1)
}

// UpdateHLCV updates the indicator with the given high, low, close, and volume values.
func (s *AdvanceDeclineOscillator) UpdateHLCV(high, low, close, volume float64) float64 {
	if math.IsNaN(high) || math.IsNaN(low) || math.IsNaN(close) || math.IsNaN(volume) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	// Compute cumulative AD.
	temp := high - low
	if temp > 0 {
		s.ad += ((close - low) - (high - close)) / temp * volume
	}

	// Feed AD to both MAs.
	fast := s.fastMA.Update(s.ad)
	slow := s.slowMA.Update(s.ad)
	s.primed = s.fastMA.IsPrimed() && s.slowMA.IsPrimed()

	if math.IsNaN(fast) || math.IsNaN(slow) {
		s.value = math.NaN()

		return s.value
	}

	s.value = fast - slow

	return s.value
}

// UpdateBar updates the indicator given the next bar sample.
// This shadows LineIndicator.UpdateBar to extract HLCV from the bar.
func (s *AdvanceDeclineOscillator) UpdateBar(sample *entities.Bar) core.Output {
	value := s.UpdateHLCV(sample.High, sample.Low, sample.Close, sample.Volume)

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: value}

	return output
}
